docs/branching-and-tagging.md

```diff
@@ -0,0 +1,218 @@
+
+Branches and tags have a maximum reference age property 
+which control when the reference to the snapshot itself should be expired. Branches have retention properties which define the minimum number of snapshots to retain on a branch as well as the maximum age of individual snapshots to retain on the branch. These properties are used when the expireSnapshots procedure is run. For details on the algorithm for expireSnapshots, refer to the [spec](../../../spec/#snapshot-retention-policy).
+
+## Usage 
+
+### Iceberg Java Library
+
+The Iceberg Java library can be used for performing operations on branches and tags.
+
+#### Creating branches and tags
+
+New branches and tags can be created via the Java library's ManageSnapshots API. 
+
+```java
+import org.apache.iceberg.Table;
+
+Table table = catalog.loadTable(...);
+
+/* Create a branch test-branch which is retained for 1 week, and the latest 2 snapshots on test-branch will always be retained. 
+Snapshots on test-branch which are created within the last hour will also be retained. */
+
+String branch = "test-branch";
+table.manageSnapshots()
+    .createBranch(branch, 3)
+    .setMinSnapshotsToKeep(branch, 2)
+    .setMaxSnapshotAgeMs(branch, 3600000)
+    .setMaxRefAgeMs(branch, 604800000)
+    .commit();
+
+// Create a tag historical-tag at snapshot 10 which is retained for a day
+String tag = "historical-tag"
+table.manageSnapshots()
+    .createTag(tag, 10)
+    .setMaxRefAgeMs(tag, 86400000)
+    .commit();
+```
+
+#### Writing to branches
+
+Writing to a branch can be performed by specifying `toBranch` in the operation. For the full list refer to [UpdateOperations](../../java/api/#update-operations). 
+```java
+
+// Append FILE_A to branch test-branch 
+String branch = "test-branch";
+
+table.newAppend()
+    .appendFile(FILE_A)
+    .toBranch(branch)
+    .commit();
+
+
+// Perform row level updates on "test-branch"
+table.newRowDelta()
+    .addRows(DATA_FILE)
+    .addDeletes(DELETES)
+    .toBranch(branch)
+    .commit();
+
+
+// Perform a rewrite operation replacing small_file_1 and small_file_2 on "test-branch" with compacted_file.
+table.newRewrite()
+    .rewriteFiles(ImmutableSet.of(small_file_1,small_file_2), ImmutableSet.of(compacted_file))
+    .toBranch(branch)
+    .commit();
+
+```
+
+#### Reading from branches and tags
+Reading from a branch or tag can be done as usual via the Table Scan API, by passing in a branch or tag in the useRef API. When a branch is passed in, the snapshot that's used is the head of the branch. Note that currently reading from a branch and specifying an asOfSnapshotId in the scan is not supported. 
```

Comment:
nit: useRef -> `useRef` so that it visually stands out when reading

Reply:
asOfSnapshotId -> `asOfSnapshotId`

Reply:
Done!